### ISSUE TITLE
API: parse full request before doing things and use transactions

### ISSUE THREAD

**Issue by pieterlexis** [PowerDNS/pdns#6053]
- Program: Authoritative, Recursor, dnsdist
 - Issue type: Feature Request

### Short description
Right now, when a POST or PUT is done through the API, we check each element seperately and perform the associated action (e.g. in `updateDomainSettingsFromDocument`), leading to possible inconsistencies when one element (e.g. "soa_edit_api") is set successfully but e.g. "api_rectify" throws a `JsonException`.

Secondly, a request should fully fail or be fully successful. So every API request that changes more than one thing should probably be wrapped inside a transaction.

**Comment by zeha** [comment]
Yup. Best if everything is checked before, and then all uncheckable things happen inside one transaction.

**Comment by Habbie** [comment]
#7891 improved this situation; we need to check if it fully fixed it.
